LayoutRenderer - Introduced IStringValueRenderer-interface to skip extra string allocation, removed IRawValue from non-ThreadAgnostic layout renderers

snakefoot opened the pull request:
SimpleLayout with single LayoutRenderer can sometimes convert directly to string without first writing to StringBuilder and then allocating new string:

The most important once are ${logger} and ${level} but there are also settings lookups ${gdc}, ${mdc}, ${mdlc}:

- LevelLayoutRenderer
- LoggerNameLayoutRenderer
- EventPropertiesLayoutRenderer
- GdcLayoutRenderer
- MdlcLayoutRenderer
- MdcLayoutRenderer

Others also included in the optimization:

- TraceActivityIdLayoutRenderer
- ThreadNameLayoutRenderer
- MachineNameLayoutRenderer
- GuidLayoutRenderer
- EnvironmentLayoutRenderer
- DateLayoutRenderer
- CurrentDirLayoutRenderer
- CallSiteFileNameLayoutRenderer

This optimizes Target.RenderLogEvent and Layout.Render for simple layouts that just contains a single dynamic value (based on the above layoutrenderers)

snakefoot commented:
Changed behavior for SimpleLayout.TryGetRawValue() so it checks if Layout.Precalculate has been performed, and uses the cached value instead of the raw value (if available).

304NotModified commented:
> Changed behavior for SimpleLayout.TryGetRawValue(), so it checks if Layout.Precalculate has been performed, and uses the cached value instead of the raw value (if available). Is this a good fallback

I think this is OK, but not sure. The cache is per logevent, isn't? So then I think it's OK

snakefoot commented:
> I think this is OK, but not sure. The cache is per logevent, isn't? So then I think it's OK

The logic calling TryGetRawValue() just need to be aware that it will suddenly return a string-object instead of ex. a DateTime-object, when the DatabaseTarget is wrapped in Async.

304NotModified commented:
> The logic calling TryGetRawValue() just need to be aware that it will suddenly return a formatted string-object instead of ex. a DateTime-object, when the DatabaseTarget is wrapped in Async.

Ow there is the TryGetRawValue not built for. Maybe we should wrap it in a "TryGetRawValueOrFallback", so the implementor could use the parse fallback

snakefoot commented:
> Ow there is the TryGetRawValue not built for. Maybe we should wrap it in a "TryGetRawValueOrFallback", so the implementor could use the parse fallback

Now changed so SimpleLayout will return false in TryGetValue when raw-value comes from MutableUnsafe-Layout and the value is not immutable. Also removed IRawValue-interface from all LayoutRenderers that are not ThreadAgnostic.

snakefoot commented:
@304NotModified Ready for review. Now IRawValue is only available when thread-agnostic to make it more consistent when using using `targets async="true"`

304NotModified commented on src/NLog/Layouts/SimpleLayout.cs:
```diff
@@ -138,6 +141,11 @@ public string Text
+        /// <summary>
+        /// Is the message a simple formatted string? (Can skip StringBuilder)
+        /// </summary>
+        internal bool IsSimpleString => _stringValueRenderer != null;
```
Was thinking of `IsSingleStringText`. "simple" is a bit ambiguous here.

304NotModified replied:
I think the name of `_stringValueRenderer ` should be in-sync with "simpleString" (or the new name)

snakefoot replied:
IsSingleStringText is fine with me. Will fix this

304NotModified commented on src/NLog/Internal/FormatHelper.cs:
```diff
@@ -71,9 +75,27 @@ internal static bool ObjectSupportsFormatProvider(object o)
             {
                 case TypeCode.String:   return false;
                 case TypeCode.Empty:    return false;
-                case TypeCode.Object:   return o is IFormattable;
```
maybe we should rename `ObjectSupportsFormatProvider` now to `SkipFormattableToString`, or something like that.

snakefoot replied:
Ok. Will fix this.

304NotModified commented on src/NLog/Layouts/SimpleLayout.cs:
```diff
@@ -222,14 +230,28 @@ internal void SetRenderers(LayoutRenderer[] renderers, string text)
+                    }
+                    if (Renderers[0] is IRenderString stringValueRendrer)
+                    {
+                        _stringValueRenderer = stringValueRendrer;
```
I like this optimization :)

304NotModified commented on src/NLog/Internal/IRenderString.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// Supports rendering of string value without allocation (or seldom)
+    /// </summary>
+    interface IRenderString
```
proposal: IStringRenderable, as class/interfaces should not contain verbs

snakefoot replied:
Ok. Will fix this

snakefoot replied:
Renamed to ISimpleStringText

snakefoot replied:
Maybe ISimpleStringValue is better than ISimpleStringText ? (Similar to IRawValue)

304NotModified commented on src/NLog/LayoutRenderers/GuidLayoutRenderer.cs:
```diff
@@ -44,22 +44,14 @@ namespace NLog.LayoutRenderers
     /// </summary>
     [LayoutRenderer("guid")]
     [ThreadSafe]
-    public class GuidLayoutRenderer : LayoutRenderer, IRawValue
```
(e.g. uniqueidentifier  type in SQL Server)

snakefoot replied:
I guess it can be made ThreadAgnostic. Not like the Guid will be much different being created on Timer-thread.

304NotModified replied:
I think this is also ThreadAgnostic, because it creates always a new guid  / or it uses only values of the logevent.

snakefoot replied:
Will fix this.

snakefoot commented on src/NLog/LayoutRenderers/EventPropertiesLayoutRenderer.cs:
```diff
@@ -90,19 +78,34 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
         {
-            GetValue(logEvent, out var value);
-            return value;
+            if (Format != "@")
```
Been there since NLog 4.5. Means serialize as JSON.

snakefoot commented on src/NLog/LayoutRenderers/LoggerNameLayoutRenderer.cs:
```diff
@@ -52,29 +53,33 @@ public class LoggerNameLayoutRenderer : LayoutRenderer
+            if (ShortName)
             {
-                builder.Append(logEvent.LoggerName);
+                int lastDot = logEvent.LoggerName?.LastIndexOf('.') ?? -1;
```
It is duplicated for performance reasons

304NotModified replied:
Could you please add a code comment for it then? thx!

snakefoot replied:
Ok. Will fix this

snakefoot commented on src/NLog/Targets/Target.cs:
```diff
@@ -744,6 +746,11 @@ protected string RenderLogEvent(Layout layout, LogEventInfo logEvent)
                     }
                 }
 
+                if (simpleLayout != null && simpleLayout.IsSimpleString)
```
Because IsFixedText means it is threadagnostic. But IsSimpleString is not guaranteed to be threadagnostic. So have to check precalculate-cache first.

snakefoot commented on tests/NLog.UnitTests/Layouts/SimpleLayoutOutputTests.cs:
```diff
@@ -188,7 +206,23 @@ public void TryGetRawValue_MultipleLayoutRender_ShouldGiveNullRawValue()
         }
 
+        [Fact]
+        public void TryGetRawValue_ImmutableLayoutRender_ShouldGiveRawValue()
```
Will fix it so it becomes Guid.

snakefoot commented:
@304NotModified Implemented all suggested changes

snakefoot commented on src/NLog/LayoutRenderers/MdcLayoutRenderer.cs:
```diff
@@ -58,30 +58,32 @@ public class MdcLayoutRenderer : LayoutRenderer, IRawValue
+        private string GetStringValue(LogEventInfo logEvent)
         {
-            return GetValue();
+            if (Format != "@")
```
Placed the magic string-value on ValueFormatter.

304NotModified commented on src/NLog/Internal/ISimpleStringText.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// Supports rendering of string value without allocation (or seldom)
+    /// </summary>
+    interface ISimpleStringText
```
I was also thinking about this, but not sure if this is a good name. 

- IsSimpleStringText is only when there is one ISimpleStringText, so then that property should be renamed to `IsSingleSimpleText`... 

Another option is `IFormattedTextRendererable` (better name is desirable)

snakefoot replied:
ISimpleStringValue ? (Similar to IRawValue)

304NotModified replied:
I think that a good idea. :)

snakefoot replied:
Maybe a combination IStringValueRenderer

snakefoot replied:
Maybe even consider changing IRawValue to IRawValueRenderer

304NotModified replied:
I don't prefer "renderer" in the interface, it doesn't render IMO (especially not the rawValue).

But luckily they are both internal, so we could change it later.

304NotModified commented on src/NLog/Internal/ISimpleStringText.cs:
```diff
@@ -0,0 +1,48 @@
+namespace NLog.Internal
+{
+    /// <summary>
+    /// Supports rendering of string value without allocation (or seldom)
```
Maybe we should explain "seldom" (`<remarks>`?). When should we implement this interface? When is it seldom? e.g. in the happy path no allocations? Or not many allocations?

snakefoot replied:
Seldom means Guid.ToString() or DateTime.ToString. I can change the comment to "limited or no allocations."

snakefoot replied:
Updated the xml-comment

304NotModified commented:
Thanks for the changes!
